src/wasm/wat-parser-internal.h

```diff
@@ -0,0 +1,521 @@
+      ctx.take(*lexed);
+    } else if (auto lexed = comment(ctx.next())) {
+      ctx.take(*lexed);
+    } else {
+      break;
+    }
+  }
+  return ctx.lexed();
+}
+
+bool LexCtx::can_finish() const {
+  // Logically we want to check for eof, parens, and space. But we don't
+  // actually want to parse more than a couple characters of space, so check for
+  // individual space chars or comment starts instead.
+  return empty() || lparen(next()) || rparen(next()) || spacechar(next()) ||
+         starts_with(";;"sv);
+}
+
+// num   ::= d:digit => d
+//         |  n:num '_'? d:digit => 10*n + d
+// digit ::= '0' => 0 | ... | '9' => 9
+std::optional<LexIntResult> num(std::string_view in) {
+  LexIntCtx ctx(in);
+  if (!ctx.take_digit()) {
+    return {};
+  }
+  while (true) {
+    bool under = ctx.take_prefix("_"sv);
+    if (!ctx.take_digit()) {
+      if (!under) {
+        return ctx.lexed();
+      }
+      return {};
+    }
+  }
+}
+
+// hexnum   ::= h:hexdigit => h
+//            | n:hexnum '_'? h:hexdigit => 16*n + h
+// hexdigit ::= d:digit => d
+//            | 'A' => 10 | ... | 'F' => 15
+//            | 'a' => 10 | ... | 'f' => 15
+std::optional<LexIntResult> hexnum(std::string_view in) {
+  LexIntCtx ctx(in);
+  if (!ctx.take_hexdigit()) {
+    return {};
+  }
+  while (true) {
+    bool under = ctx.take_prefix("_"sv);
+    if (!ctx.take_hexdigit()) {
+      if (!under) {
+        return ctx.lexed();
+      }
+      return {};
+    }
+  }
+}
+
+// uN ::= n:num         => n (if n < 2^N)
+//      | '0x' n:hexnum => n (if n < 2^N)
+// sN ::= s:sign n:num         => [s]n (if -2^(N-1) <= [s]n < 2^(N-1))
+//      | s:sign '0x' n:hexnum => [s]n (if -2^(N-1) <= [s]n < 2^(N-1))
+// sign ::= {} => + | '+' => + | '-' => -
+//
+// Note: Defer bounds and sign checking until we know what kind of integer we
+// expect.
+std::optional<LexIntResult> integer(std::string_view in) {
+  LexIntCtx ctx(in);
+  ctx.take_sign();
+  if (ctx.take_prefix("0x"sv)) {
+    if (auto lexed = hexnum(ctx.next())) {
+      ctx.take(*lexed);
+      if (ctx.can_finish()) {
+        return ctx.lexed();
+      }
+    }
+    // TODO: Add error production for unrecognized hexnum.
+    return {};
+  }
+  if (auto lexed = num(ctx.next())) {
+    ctx.take(*lexed);
+    if (ctx.can_finish()) {
+      return ctx.lexed();
+    }
+  }
+  return {};
+}
+
+// ======
+// Tokens
+// ======
+
+struct LParenTok {
+  friend std::ostream& operator<<(std::ostream& os, const LParenTok&) {
+    return os << "'('";
+  }
+};
+
+struct RParenTok {
+  friend std::ostream& operator<<(std::ostream& os, const RParenTok&) {
+    return os << "')'";
+  }
+};
+
+struct IntTok {
+  uint64_t n;
+  bool hasSign;
+  friend std::ostream& operator<<(std::ostream& os, const IntTok& tok) {
+    return os << tok.n << " sign: " << tok.hasSign;
+  }
+};
+
+struct Token {
+  using Data = std::variant<LParenTok, RParenTok, IntTok>;
+
+  std::string_view span;
+  Data data;
```

Comment:
> * Why should `Token` have both `span` and `data`? I think basically they contain the same info.. Isn't one of these kind of redundant?

Tokens need their spans, their kinds, and sometimes additional additional data like numbers. With the current factorization of the code, `Token` contains the span and the kind (via the variant) and `*Tok` contain the optional additional data. There's no redundancy here because the `*Tok` classes do no contain spans.

I experimented with duplicating the `span` into each of the `*Tok` classes to make them more independently useful and to make `Token` a simple variant wrapper, but the codegen for getting the span from a `Token` was bad because the optimizer could not see through the `variant` representation.

> * Why should `***Tok` classes and this `Token` class be separate classes?

`Token` is the union of many different kinds of tokens, each of which is represented with a separate `*Tok` class. Another way of representing this would have been to have `Token` be the common superclass of all the `*Tok` types, but then I would have needed to add an explicit downcasting mechanism. Using `std::variant` seems like the more modern way to do this, but maybe I'm just biased because I wish I were writing Rust :crab:

Reply:
Ah, I forgot to mention that another goal of the experiment I described was to combine the `Lex*Result` classes with the `*Tok` classes, since you're right that those are very similar and it would be nice to deduplicate them. Maybe it was a premature optimization to look at the resulting codegen, but I think it's ok. The potential code simplification wasn't huge.

Reply:
Thanks for the explanation! I wasn't suggesting deduplicating them for optimization's sake; I was mostly trying to understand each class' purpose, and they looked very similar. I'm still not sure if we need `LexResult` given that we have `***Tok`s, but I guess following patches may clarify them. I guess we will add more `***Tok` classes, right? Are we adding more `Lex***Result` classes as well?

Reply:
Yes, we will add more `***Tok` like `IdTok`, `KeywordTok`, and `StringTok`. I think the only additional `LexResult` we will need is `StringLexResult`.